File: app/models/assignment.rb
class Assignment < ActiveRecord::Base
  validates :due_date, :assigned_date, :worksheet, presence: true
  belongs_to :classroom
  belongs_to :worksheet

  validate :validate_assigned_period, :due_in_future, :assigned_in_future

  def validate_assigned_period
    if assigned_date && due_date && assigned_date >= due_date
        errors.add(:due_date, "must be later than the assigned_date")
    end
  end

  def due_in_future
    if due_date && due_date < 1.minute.ago
        errors.add(:due_date, "must be in the future")
    end
  end

  def assigned_in_future
    if assigned_date && assigned_date < 1.minute.ago
        errors.add(:assigned_date, "must be in the future")
    end
  end
end

File: app/assets/javascripts/models/assignment.js
App.Assignment = DS.Model.extend({
  classroom: DS.belongsTo('classroom', {async: true}),
  worksheet: DS.belongsTo('worksheet'),
  dueDate: DS.attr('date'),
  assignedDate: DS.attr('date')
})

App.Assignment.FIXTURES = [{
  id: 1,
  classroom: 1,
  worksheet: 1,
  dueDate: new Date('2015'),
  assignedDate: new Date('2015')
},{
  id: 2,
  classroom: 1,
  worksheet: 1,
  dueDate: new Date('2015'),
  assignedDate: new Date('2013')
},{
  id: 3,
  classroom: 2,
  worksheet: 1,
  dueDate: new Date('2013'),
  assignedDate: new Date('2013')
}]
